customer-models/src/main/java/org/powertac/customer/evcharger/DemandSampler.java

```diff
@@ -0,0 +1,256 @@
+      for (int j = 0; j <= maxChargerHours; j++) {
+        cohortChargerHoursHistogram.get(i).put(j, 0);
+      }
+      cohortVehicleSum.put(i, 0);
+      cohortEnergySum.put(i, 0.0);
+    }
+
+    // Now, we fill in the the values for the charger hour histograms, vehicle
+    // sum and energy sum of each cohort.
+    for (double[] horizonEnergyTuple: horizonEnergyTuples) {
+      TreeMap<Integer, Integer> histogram = cohortChargerHoursHistogram.get((int) horizonEnergyTuple[0]);
+      histogram.merge((int) (horizonEnergyTuple[1] / chargerCapacity), 1, Integer::sum);
+      cohortChargerHoursHistogram.put((int) horizonEnergyTuple[0], histogram);
+
+      cohortVehicleSum.merge((int) horizonEnergyTuple[0], 1, Integer::sum);
+      cohortEnergySum.merge((int) horizonEnergyTuple[0], horizonEnergyTuple[1], Double::sum);
+    }
+
+    return cohortVehicleSum.keySet().stream()
+            .map(horizon -> new DemandElement(horizon, cohortVehicleSum.get(horizon), cohortEnergySum.get(horizon),
+                                              cohortChargerHoursHistogram.get(horizon).values().stream()
+                                                      .mapToDouble(Integer::doubleValue).toArray()))
+            .collect(Collectors.toList());
+  }
+
+  /**
+   * Sample the number of new plug-ins for a given hour of day. Returns the
+   * result as a double to allow for better accuracy in the simulation.
+   *
+   * @param hod
+   *          Current hour of day
+   * @param popSize
+   *          Population size of chargers
+   * @return Number of new plug-ins for the current timeslot
+   */
+  double sampleNewPlugins (final int hod, final int popSize)
+  {
+    double result = pluginProbability.density(new double[] { hod }) * popSize;
+    final NormalDistribution gaussianNoise = new NormalDistribution(0, result * 0.1);
+    if (seed != null) {
+      gaussianNoise.reseedRandomGenerator(seed);
+    }
+    result += gaussianNoise.sample();
+    return Math.max(0, result);
+  }
+
+  /**
+   * Sample n (horizon, energy demand) tuples.
+   * 
+   * @param n
+   *          The number of samples to be retrieved. This is typically given by
+   *          {@link DemandSampler#sampleNewPlugins}.
+   * @param hod
+   *          The current hour of day as integer. Must be in the interval
+   *          [0, 23].
+   * @throws IllegalArgumentException
+   *           If no distribution is found for the given hour of day.
+   * @return Tuples of horizon and energy demand. The return dimension is n x 2.
+   *         Samples are guaranteed to be larger or equal to zero.
+   */
+  double[][] sampleHorizonEnergyTuples (final int n, final int hod)
+  {
+    MixtureMultivariateNormalDistribution condDist = condHorizonDemandProbabilities.get("hod" + hod);
+    if (condDist == null) {
+      throw new IllegalArgumentException(String.format("Cannot find distribution for provided hour of day %s.", hod));
```

Comment:
Do we log exception or do we need an additional log call here?

Reply:
If we throw an exception and don't catch it, it will be caught by CompetitionControlService.SimRunner and will crash the server. If this is a fatal exception for the model, then logging the error and disabling the model might make more sense.